src/kernel/chainstatemanager_opts.h

```diff
@@ -24,6 +25,8 @@ namespace kernel {
 struct ChainstateManagerOpts {
     const CChainParams& chainparams;
     const std::function<NodeClock::time_point()> adjusted_time_callback{nullptr};
+    /** Block hash whose ancestors we will assume to have valid scripts without checking them. */
+    uint256 assumed_valid_block;
```

Comment:
I think it would be best if the libbitcoinkernel context initialization took care of initializing the value instead of turning all validation globals into tristates and all places where they are used, rely on ternary operators.

Thus, I think it is best to keep this as `uint256`.

Reply:
Ok, I can see how `std::optional` is helpful. Though, I'd prefer to at least wrap the ternary operators into a member function. Maybe I should add a `const uint256&  ChainstateManager::AssumedValidBlock() const { return m_settings.assumed_valid_block ? *m_settings.assumed_valid_block : GetConsensus().defaultAssumeValid; }`?

Reply:
FYI when @dongcarl and I discussed the generic pattern to use for overrides a few months back, I played around with (and liked) communicating via `std::optional` as well, like here: https://github.com/theuni/bitcoin/commit/7fad1352fb1aea12c41dfb8fccd8b220a58b694d


> Though, I'd prefer to at least wrap the ternary operators into a member function. Maybe I should add a `const uint256& ChainstateManager::AssumedValidBlock() const { return m_settings.assumed_valid_block ? *m_settings.assumed_valid_block : GetConsensus().defaultAssumeValid; }`?


+1 for the convenient short-hand.

Reply:
> I'd prefer to at least wrap the ternary operators into a member function. Maybe I should add a `const uint256& ChainstateManager::AssumedValidBlock() const { return m_settings.assumed_valid_block ? *m_settings.assumed_valid_block : GetConsensus().defaultAssumeValid; }`?

Yes that's what I meant by "define accessor methods on chain or block classes to combine validation options & chain params as needed"

Reply:
Hmm, on a second though I am wondering if this is a good idea. `std::optional` has cmp functions that accept `std::nullopt` (https://en.cppreference.com/w/cpp/utility/optional/operator_cmp). Thus, it wouldn't be a compile failure to not use the accessor method.

It might be better to "flatten" the options while storing them in m_options in the chainstatemanager constructor.

Reply:
I'm not sure what issue you're referring to with the comparison functions. I don't think you ever need to call them

Reply:
The compare functions are used to check the minimum chain work. But it may also happen for other types that can be compared. For example, if there is an `optional<int> num`, with a chainparams fallback. A dev might accidentally type `if (m_options.num > 4)` instead of `if (Num() > 4)`, thus not using the fallback. Since the code compiles (and looks plausible), the would likely not be caught during review.

Reply:
That's fair, but what you are describing sounds like a reason to avoid std::optional class entirely in sensitive validation or wallet code for types that can be compared. It would apply just as much to internal state as external options (maybe even more) and would probably suggest that we should stop using std::optional in existing code.

I think it's a reasonable concern, just not a concern that's unique to this situation or extra compelling here. One way to address it would be to use `std::variant` instead of `std::optional` like:

```c++
struct UseChainDefault {};

struct Options {
  std::variant<uint256, UseChainDefault> min_work = UseChainDefault{};
};

uint256 GetMinWork()
{
    return std::holds_alternative<UseChainDefault>(m_options.min_work)
        ? m_params.GetConsensus().nMinimumChainWork
        : std::get<uint256>(m_options.min_work);
};
```

Reply:
I am actually thinking that we should disallow the compare operator on `std::optional`. However, that is a separate topic.

Here, I'd simply "flatten" the `optional`s in the chainstatemanager constructor. Meaning the type will remain `std::optional<bla>`, however the optional is guaranteed to hold a value (either the previous value or a copy of the chain default value)

Reply:
Oh, I wasn't sure what you meant with flattening, but that's a reasonable approach. I don't love it because I think immutability is nice, so data is passed around verbatim, and you can set a value in one part of the code, and read it in another part of the code, and not have to worry about hidden code in the middle tweaking things. But this is nitpicking. The main goal is just to have an options struct that uses existing bitcoind default values and doesn't introduce a new set of default behaviors.

Reply:
Got the summon from @theuni, skimmed the comments here... I don't particularly care either way but I think an alternative is to discard the concept of `ChainstateManagerOpts` when we construct `ChainstateManger`:

```cpp
struct ChainstateManagerOpts {
    const CChainParams& chainparams;
...
    std::optional<uint256> assumed_valid_block;
...
}
```

And have the `ChainstateManager` resolve this logic in its constructor, but without storing a `m_options`:

```cpp
ChainstateManager::ChainstateManager(Options options)
    : m_effective_assume_valid_block{
          options.assumed_valid_block.value_or(options.chainparams.GetConsensus().defaultAssumeValid)
      }, ...
```

That way past the `ChainstateManager` ctor boundary there's no confusion about what to use. 

These `*Opts` structs don't really need to be (and in a lot of cases shouldn't be) stored as-is in their corresponding classes , classes can do smart things with them in their constructors and discard the structure.

Reply:
> These *Opts structs don't really need to be (and in a lot of cases shouldn't be) stored as-is in their corresponding classes

I disagree with this. Yes, it is possible to come up with different representations of the same options, transforming them from one state to another, with different representations optimized to take up less space or avoid recomputing values, or provide more type safety.

But I think this overcomplicates things, which is why I added the `m_options` struct in #25905 and am encouraging #25862 as an alternative to #25623. IMO, it is best to choose to choose **one** reasonable representation of options, stop transforming them from one state to another, come to peace with the fact that no single representation will be the most convenient or safest or most efficient form for every stage of the application, but that the simplicity of having one representation is better than the alternative of introducing boilerplate code in lots of places to copy or transform option values.

I think this is true for most options, allowing that there could be some exceptions.

Reply:
Fixed